rpc/handlers.go

```diff
@@ -1038,3 +1038,48 @@ func (h *Handler) Call(call FunctionCall, id BlockID) ([]*felt.Felt, *jsonrpc.Er
 	}
 	return res, nil
 }
+
+func (h *Handler) EstimateFee(broadcastedTxns []BroadcastedTransaction, id BlockID) ([]*FeeEstimate, *jsonrpc.Error) {
+	var estimates []*FeeEstimate
+	state, closer, err := h.stateByBlockID(&id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer func() {
+		if closeErr := closer(); closeErr != nil {
+			h.log.Warnw("Failed to close state in starknet_call", "err", closeErr)
+		}
+	}()
+
+	header, err := h.blockHeaderByID(&id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	if header.GasPrice == nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, "no gas price history for the given block")
+	}
+
+	for idx := range broadcastedTxns {
+		txn, declaredClass, err := adaptBroadcastedTransaction(&broadcastedTxns[idx])
+		if err != nil {
+			return nil, jsonrpc.Err(jsonrpc.InvalidParams, err.Error())
+		}
+
+		gasConsumed, err := vm.Execute(txn, declaredClass, header.Number, header.Timestamp, state, h.network)
```

Comment:
From the spec:

> a sequence of transactions to estimate, running each transaction on the state resulting from applying all the previous ones

Do we update the state somewhere? Since `callContext` uses a `state.Reader`, it doesn't seem that we do. Based on the [blockifier implementation](https://github.com/starkware-libs/blockifier/blob/a3a133af7eaa07197caa90766db11edacf8f46f6/crates/blockifier/src/transaction/transactions.rs#L30-L48), it seems that the state is only modified on their side but never passed to us?

Reply:
> Do we update the state somewhere

Nope, nice catch.